### PR TITLE
ARROW-16990: [C++] Typos/copy-paste mistakes in Substrait API docstrings

### PR DESCRIPTION
**Pull request by jvanstraten**
Reformatted and fixed the docstrings for the Substrait API. No functional changes.

### CODE REVIEW

File: cpp/src/arrow/engine/substrait/serde.h [hunk 1]
```diff
@@ -39,38 +39,38 @@ namespace engine {
...
     const Buffer& buf, const ConsumerFactory& consumer_factory,
     const ExtensionIdRegistry* registry = NULLPTR, ExtensionSet* ext_set_out = NULLPTR);
 
-/// \brief Deserializes a single-relation Substrait Plan message to an execution plan
+/// \brief Deserializes a single-relation Substrait Plan message to an execution plan.
 ///
-/// The output of each top-level Substrait relation will be sent to a caller supplied
-/// consumer function provided by consumer_factory
+/// The output of each top-level Substrait relation will be sent to the given consumer
+/// node.
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde.h:63]:
```suggestion
/// The output of each top-level Substrait relation will be sent to the given consumer.
```
Minor nit: A `SinkNodeConsumer` isn't actually an exec node.  An exec node has to worry about various lifecycle concerns but a consumer just receives batches.

File: cpp/src/arrow/engine/substrait/serde.h [hunk 2]
```diff
@@ -79,150 +79,151 @@ Result<compute::ExecPlan> DeserializePlan(
...
 /// The output of the single Substrait relation will be written to a filesystem.
 /// `write_options_factory` can be used to control write behavior.
 ///
 /// \param[in] buf a buffer containing the protobuf serialization of a Substrait Plan
-/// message
-/// \param[in] write_options write options of a node consuming the batches produced by
-/// each toplevel Substrait relation
+/// message.
+/// \param[in] write_options write options for the node that will consume the batches
+/// produced by the toplevel Substrait relation.
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde.h:108]:
```suggestion
/// produced by the top level Substrait relation.
```
Or top-level?

**Review by westonpace — APPROVED** [review]
Just some minor thoughts.  Thanks for the cleanup

File: cpp/src/arrow/engine/substrait/serde.h (hunk 2 above)
Comment (jvanstraten) [reply cpp/src/arrow/engine/substrait/serde.h:108]: That would be "top-level" (it's a compound adjective).

File: cpp/src/arrow/engine/substrait/serde.h [hunk 3]
```diff
@@ -36,193 +36,193 @@ namespace arrow {
 namespace engine {
 
 /// Factory function type for generating the node that consumes the batches produced by
-/// each toplevel Substrait relation when deserializing a Substrait Plan.
+/// each top-level Substrait relation when deserializing a Substrait Plan.
 using ConsumerFactory = std::function<std::shared_ptr<compute::SinkNodeConsumer>()>;
 
-/// \brief Deserializes a Substrait Plan message to a list of ExecNode declarations
+/// \brief Deserializes a Substrait Plan message to a list of ExecNode declarations.
```
Comment (pitrou) [inline cpp/src/arrow/engine/substrait/serde.h:42]: While we are at it, can we follow the convention of using the infinitive in function summaries.
```suggestion
/// \brief Deserialize a Substrait Plan message to a list of ExecNode declarations.
```

File: cpp/src/arrow/engine/substrait/serde.h [hunk 4]
```diff
@@ -36,193 +36,193 @@ namespace arrow {
...
-/// \return success if equivalent, failure if not
+/// \param[in] message_name the name of the Substrait message type to check.
+/// \param[in] l_buf buffer containing the first protobuf serialization to compare.
+/// \param[in] r_buf buffer containing the second protobuf serialization to compare.
+/// \return success if equivalent, failure if not.
 ARROW_ENGINE_EXPORT
 Status CheckMessagesEquivalent(util::string_view message_name, const Buffer& l_buf,
                                const Buffer& r_buf);
 
 /// \brief Utility function to convert a JSON serialization of a Substrait message to
```
Comment (pitrou) [inline cpp/src/arrow/engine/substrait/serde.h:209]: Or if we want to keep it concise
```suggestion
/// \brief Convert a JSON serialization of a Substrait message to
```

Comment (jvanstraten) [reply cpp/src/arrow/engine/substrait/serde.h]: 6d61c85

File: cpp/src/arrow/engine/substrait/serde.h (hunk 3 above)
Comment (jvanstraten) [reply cpp/src/arrow/engine/substrait/serde.h]: 6d61c85

**Review by pitrou — APPROVED** [review]
+1. @westonpace Would you like to do the final merge?

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Pull request closed without merging**